model/stats/presto_query.go

```diff
@@ -0,0 +1,178 @@
+package stats
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"text/template"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+)
+
+const (
+	testStatsQueryTemplateString = `
+WITH matched_and_aggregated AS (
+    SELECT  test_name,{{ if not .GroupByTest }}
+            task_name,{{ end }}
+	    {{ if .GroupDays }}MIN(task_create_iso) AS task_create_iso{{ else }}task_create_iso{{ end }},
+	    SUM(num_pass) AS num_pass,
+	    SUM(num_fail) AS num_fail,
+	    SUM(total_pass_duration_ns) AS total_pass_duration_ns
+    FROM test_stats_daily
+    WHERE project = ?
+    AND   variant = ?{{ if .TaskName }}
+    AND   task_name = ?{{ end }}{{ if .TestName }}
+    AND   test_name = ?{{ end }}
+    AND   request_type IN (?{{ range slice .Requesters 1}}, ?{{end}})
+    AND   task_create_iso BETWEEN ? AND ?
+    GROUP BY test_name{{ if not .GroupByTest }}, task_name{{ end }}{{ if not .GroupDays }}, task_create_iso{{ end }}
+    ORDER BY task_create_iso{{ if .SortDesc }} DESC{{ end }},{{ if not .GroupByTest }} task_name ASC,{{ end }} test_name ASC
+    OFFSET ?
+    LIMIT ?
+)
+SELECT  test_name,{{ if not .GroupByTest }}
+        task_name,{{ end }}
+	date(task_create_iso) AS date,
+        num_pass,
+        num_fail,
+        IF(num_pass <> 0, total_pass_duration_ns/num_pass, 0) AS average_duration
+FROM matched_and_aggregated
+`
+	dateFormat = "2006-01-02"
+)
+
+var testStatsQueryTemplate = template.Must(template.New("test_stats_api").Parse(testStatsQueryTemplateString))
+
+// PrestoTestStatsFilter represents search and aggregation parameters when
+// querying test statistics in Presto.
+type PrestoTestStatsFilter struct {
+	Project    string
+	Variant    string
+	TaskName   string
+	TestName   string
+	Requesters []string
+
+	AfterDate  time.Time
+	BeforeDate time.Time
+
+	Offset   int
+	Limit    int
+	SortDesc bool
+
+	GroupByTest bool
+	GroupDays   bool
+
+	DB *sql.DB
+}
+
+// Validate checks that the filter is valid and, if necessary, populates
+// default values.
+func (f *PrestoTestStatsFilter) Validate() error {
+	catcher := grip.NewBasicCatcher()
+
+	catcher.NewWhen(f.Project == "", "must specify a project")
+	catcher.NewWhen(f.Variant == "", "must specify a variant")
+	catcher.NewWhen(f.TaskName == "" && f.TestName == "", "must specify a task name and/or test name")
+	if len(f.Requesters) == 0 {
+		f.Requesters = []string{evergreen.RepotrackerVersionRequester}
+	}
+
+	today := utility.GetUTCDay(time.Now())
+	f.AfterDate = utility.GetUTCDay(f.AfterDate)
+	f.BeforeDate = utility.GetUTCDay(f.BeforeDate)
+	catcher.NewWhen(f.AfterDate.After(f.BeforeDate), "before date must be greater than or equal to after date")
+	catcher.NewWhen(today.Sub(f.AfterDate) > 180*24*time.Hour, "must specify an after date within 180 days from today")
```

Comment:
Yeah, there isn't really a TTL in Presto data for this but we want to constrain inputs and this is consistent with the legacy route. I don't think its worth making a constant for, this will go away eventually anyways.

Reply:
I guess this is again a question of "how soon is eventually" - if it's a specific project/time, I'm fine with leaving it.